src/goModules.ts

```diff
@@ -1,94 +1,84 @@
-import { getBinPath, getGoVersion, sendTelemetryEvent, getToolsEnvVars, getCurrentGoPath } from './util';
+import { getBinPath, getGoVersion, sendTelemetryEvent, getToolsEnvVars } from './util';
 import path = require('path');
 import cp = require('child_process');
 import vscode = require('vscode');
 import { getFromGlobalState, updateGlobalState } from './stateUtils';
 import { installTools } from './goInstallTools';
 
-function containsModFile(folderPath: string): Promise<boolean> {
-	let goExecutable = getBinPath('go');
-	if (!goExecutable) {
-		return Promise.reject(new Error('Cannot find "go" binary. Update PATH or GOROOT appropriately.'));
-	}
-	return new Promise(resolve => {
-		cp.execFile(goExecutable, ['env', 'GOMOD'], { cwd: folderPath, env: getToolsEnvVars() }, (err, stdout) => {
-			if (err) {
-				console.warn(`Error when running go env GOMOD: ${err}`);
-				return resolve(false);
-			}
-			let [goMod] = stdout.split('\n');
-			resolve(!!goMod);
-		});
-	});
-}
 const workspaceModCache = new Map<string, boolean>();
-const packageModCache = new Map<string, boolean>();
+const folderModCache = new Map<string, string>();
 
 export function isModSupported(fileuri: vscode.Uri): Promise<boolean> {
+	return getModPath(fileuri).then(modPath => {
+		return modPath !== '';
+	});
+}
+
+export function getModPath(fileuri: vscode.Uri): Promise<string> {
+	const folderPath = path.dirname(fileuri.fsPath);
+
+	const hit = folderModCache.get(folderPath);
+	if (hit !== undefined) {
+		return Promise.resolve(hit);
+	}
+
+	workspaceModCache.forEach((v, k) => {
+		if (folderPath.startsWith(k)) {
+			folderModCache.set(folderPath, k);
+			return Promise.resolve(k);
```

Comment:
This `return` applies just to the callback passed to the `forEach` call. If you want to exit the `getModPath` function when you `folderPath.startsWith(k)` is true, then you need to use a simple `for` loop